lib/pbench/server/api/resources/query_apis/__init__.py

```diff
@@ -46,6 +46,19 @@ def __str__(self) -> str:
         return f"API {self.subclass_name} is missing schema parameters controller and/or name"
 
 
+class InvalidAPIRequest(Exception):
+    """
+    Used to report an Invalid API Request
+    """
+
+    def __init__(self, status: int, message: str, data: JSON = None):
+        self.status = status
+        self.message = message
+
+    def __str__(self) -> str:
+        return f"Invalid API request error returning {self.status} "
```

Comment:
Should this include the message value somehow?

Reply:
That's a really good point; I almost made a similar comment, but refrained. However, when we *catch* this, we do `abort(e.status, message=str(e))` ... which means the *message* is a repeat of the status along with the generic "Invalid API request" boilerplate. That's not useful at all.

We don't want either the boilerplate or the status in the `abort` `message=` key; we want the actual description, the `self.message` of the exception. First, we should probably be doing `abort(e.status, message=e.message)` instead of just string-ifying the exception at the catch point. However, the "Invalid API request" boilerplate is really a problem itself; it's not always an "invalid API request" ... it's a problem discovered in the API implementation. We don't need to expose those details to the client, regardless of the names or internal representation: we just want to report the status and message determined at the point where we raised the exception.

I'm not sure we need an `__str__` at all, but it's generally a good idea to have one. In this case, I'd suggest something more generic like f"API error '{self.message}': {self.status}".

And while I'm on the subject, I nearly suggested changing the exception name to `ApiError`... and I might as well throw that out since, again, "Invalid API request" suggests that the client did something wrong, which isn't necessarily the case. Or even `ApiAbort` since this is really our own replacement for Flask's (actually Werkzeug's) `abort` mechanism so we don't have to deal with their implicit semi-hidden exception outside the outer API implementation layer.

Reply:
It's not a `SchemaError` ... let's not reuse base classes just because it's operationally convenient, unless they're really related.

If we want to do that, though, it could be a great idea to create a new base class which all API exceptions extend, building the concept of a return `http_status` into all of them by inheritance. That could cover this new exception along with `SchemaError`, `UnauthorizedAccess`, and `PostprocessError`.